OpenRA.Mods.Common/Activities/Move/AttackMoveActivity.cs

```diff
@@ -18,31 +19,70 @@ namespace OpenRA.Mods.Common.Activities
 {
 	public class AttackMoveActivity : Activity
 	{
-		const int ScanInterval = 7;
-
+		readonly Func<Activity> getInner;
 		Activity inner;
-		int scanTicks;
+		Activity attack;
 		AutoTarget autoTarget;
+		bool moving;
 
-		public AttackMoveActivity(Actor self, Activity inner)
+		public AttackMoveActivity(Actor self, Func<Activity> getInner)
 		{
-			this.inner = inner;
+			this.getInner = getInner;
 			autoTarget = self.TraitOrDefault<AutoTarget>();
+			moving = false;
 		}
 
 		public override Activity Tick(Actor self)
 		{
-			if (autoTarget != null && --scanTicks <= 0)
+			if (IsCanceled)
 			{
-				autoTarget.ScanAndAttack(self, true);
-				scanTicks = ScanInterval;
+				if (inner != null)
+				{
+					inner = ActivityUtils.RunActivity(self, inner);
+					return this;
+				}
+
+				return NextActivity;
+			}
+
+			if (autoTarget == null)
+				return getInner();
+
+			if (attack == null)
+			{
+				var target = autoTarget.ScanForTarget(self, true);
+				if (target.Type != TargetType.Invalid)
+				{
+					if (inner != null)
+						inner.Cancel(self);
+
+					var attackBases = autoTarget.ActiveAttackBases;
+					foreach (var ab in attackBases)
+					{
+						if (attack == null)
```

Comment:
This is already in an `attack == null` branch (see line 51), so do we really need to check this again?

Reply:
It may not be null is there is more than one attackBase.

Reply:
aha, I see.  That was an obvious oversight on my part.